[Lim Kai Xuan] iP

kaixuan477 opened the pull request:
### Mel Chatbot
> Sometimes you wonder, how did you reach this place?

Mel simply helps you by remembering your tasks for you. Mel is:

- Text-based
- Straightforward to command
- ~~Simple~~ **_REALLY_ simple** and neat

What you need to get Mel:

1. Download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)
2. Double click to download
3. Command Mel to add your tasks
4. Let Mel manage your tasks for you :upside_down_face:

Of course, Mel is **FREE**!

What more features can Mel offer?

- [x] Managing tasks
- [ ] Search tasks by date/time (*coming soon*)
- [ ] Reminders (*coming soon*)

---

If you are a Java programmer, you can use Mel to practice your Java too. The `main` method looks like this:

```java
public class Mel {
    public static void main(String[] args) {
        new Mel().run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kaixuan477 committed:
level 0. change Duke to Mel

kaixuan477 committed:
level 1

kaixuan477 committed:
level 2. add Storage.java

kaixuan477 committed:
level 3. add Tasks.java and Task.java

kaixuan477 committed:
level 4. add Deadline.java, Event.java, Todo.java and remove Tasks.java

kaixuan477 committed:
A-TextUiTesting. Add test cases and results in input.txt and EXPECTED.TXT

kaixuan477 committed:
A-Exceptions. add MelException.java and TaskException.java, updated input.txt and EXPECTED.txt for test cases.

kaixuan477 committed:
Level-6. added delete functionality

kaixuan477 committed:
Update code to conform to java coding standards

kaixuan477 committed:
Implement save functionality, add ./data/tasks.txt as save file

kaixuan477 committed:
Merge branch 'branch-Level-7'

kaixuan477 committed:
implement date and time functionality, update Mel.java to limit input length

kaixuan477 committed:
Merge branch 'branch-Level-8'

kaixuan477 committed:
Add UI.java, TaskList.java and Parser.java, update files for OOP, update tests and outputs

kaixuan477 committed:
Move java files into packages, update runtest.bat and README.md

kaixuan477 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

kaixuan477 committed:
update build.gradle

kaixuan477 committed:
add test files DeadlineTest.java and ParserTest.java, and JUnit tests

kaixuan477 committed:
update build.gradle for JAR file packaging

kaixuan477 committed:
Add documentation to java files

kaixuan477 committed:
Update java files to follow java coding standard

kaixuan477 committed:
Update mel.task files to implement find functionality

kaixuan477 committed:
Merge branch 'branch-A-JavaDoc'

kaixuan477 committed:
Merge branch 'branch-A-CodingStandard'

kaixuan477 committed:
Merge branch 'branch-Level-9'

anselmlong commented on README.md:
```diff
@@ -13,12 +13,12 @@ Prerequisites: JDK 17, update Intellij to the most recent version.
-   | | | | | | | |/ / _ \
-   | |_| | |_| |   <  __/
-   |____/ \__,_|_|\_\___|
+    __    __        __  
```
I like how you reused the ASCII

anselmlong commented on gradle/wrapper/gradle-wrapper.properties:
You can gitignore the gradle files!

anselmlong commented on src/main/java/mel/exceptions/TaskException.java:
I like how you separated your exceptions!

anselmlong commented on src/main/java/mel/main/Mel.java:
```diff
@@ -0,0 +1,70 @@
+    public Mel() {
+        Storage storage = new Storage();
+        taskList = new TaskList(this, storage);
+        ui = new UI(this);
```
UI should be Ui instead, follow the naming convention [here](https://se-education.org/guides/conventions/java/intermediate.html#naming).

anselmlong commented on src/main/java/mel/main/Mel.java:
```diff
@@ -0,0 +1,70 @@
+     * @param input task input string.
+     * @see TaskList
+     */
+    public void taskAction(String input) {
```
Could use a better name here? How about something like executeTask? Function names should be verbs.

anselmlong commented on src/main/java/mel/main/Mel.java:
```diff
@@ -0,0 +1,70 @@
+     * @param str response string.
+     * @see UI
+     */
+    public void println(String str) {
```
Not sure I understand, but why do you need this function if you can just call `ui.println()`?

devivy replied:
same question

anselmlong commented on src/main/java/mel/main/Mel.java:
```diff
@@ -0,0 +1,70 @@
+     * Main method to start up Mel chatbot.
+     */
+    public static void main(String[] args) {
+        new Mel().run();
```
Clean!

anselmlong commented on src/main/java/mel/tasks/Deadline.java:
I would consider naming your tasks with the suffix Task, like DeadlineTask, to be more consistent!

devivy replied:
yes you should consider , according to me its a more consistent and intuitive way to name ur tasks

anselmlong commented on src/main/java/mel/tasks/Deadline.java:
```diff
@@ -0,0 +1,37 @@
+            String[] temp = input.split("/by ");
+            task = temp[0].split(" ", 2)[1];
+            by = super.parseDateTime(temp[1]);
+        } catch (ArrayIndexOutOfBoundsException | ParseException e) {
```
Would having separate error messages for the separate exceptions be clearer?

kaixuan477 replied:
I suppose so, although I treated them similarly here because both `ArrayIndexOutOfBoundsException` and `ParseException` would produce the same effect of incorrect formatting causing a parse error. Thus, I opted to throw the same error message with `ParseException` in this case. Thank you! 😄

anselmlong commented on src/main/java/mel/tasks/Event.java:
```diff
@@ -0,0 +1,51 @@
+            String[] temp = input.split("/from ");
+            task = temp[0];
+
+            String[] s = temp[1].split(" /to ");
```
You could name this something more intuitive, like dateSplit?

anselmlong commented on src/main/java/mel/tasks/Event.java:
```diff
@@ -0,0 +1,51 @@
+            from = super.parseDateTime(s[0]);
+            to = super.parseDateTime(s[1]);
+            if (from.isAfter(to)) {
+                throw new TaskException("event <task> "
```
Woah, love that you have date checking here!

anselmlong commented on src/main/java/mel/tasks/Event.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + task + "(from: "
+                + from.format(DateTimeFormatter.ofPattern("d LLL uuuu h.mma"))
```
Would it be better to abstract out `DateTimeFormatter.ofPattern("d LLL uuuu h.mma")` to a variable? Also, would using `String.format(%s ...)` be easier instead of adding strings?

kaixuan477 replied:
I suppose so, would abstracting `DateTimeFormatter` formatter be more convenient? Perhaps they were to the same effect.

anselmlong replied:
How would you abstract `DateTimeFormatter` yet?

anselmlong commented on src/main/java/mel/tasks/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     * @throws TaskException if input is of invalid format.
+     * @throws TaskException if task input is invalid.
+     */
+    public void taskAction(String input) throws MelException, TaskException {
```
See earlier comment about function naming!

anselmlong commented on src/main/java/mel/tasks/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     */
+    public void taskAction(String input) throws MelException, TaskException {
+        String cmd = input.split(" ", 2)[0];
+        if (Objects.equals(input, "list")) {
```
Would a switch-case statement here be better?

anselmlong commented on src/main/java/mel/tasks/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+        } else if (cmd.equals("find")) {
+            find(input);
+        } else {
+            mel.println("Mel remembers...");
```
Would it be better to just let UI handle all the printing, and declaring a UI instance here instead?

devivy replied:
yes, i feel that would be better too

anselmlong commented on src/main/java/mel/tasks/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     * @throws TaskException if input is of invalid format.
+     * @throws TaskException if task input is invalid.
+     */
+    private void add(String str) throws MelException, TaskException {
```
A clearer variable name here would be good! Something like `createTaskFromInput`?

anselmlong commented on src/main/java/mel/tasks/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     */
+    private void mark(String str) {
+        String[] temp = str.split(" ");
+        String m = temp[0];
```
More descriptive variable naming?

anselmlong commented on src/main/java/mel/utils/Parser.java:
```diff
@@ -0,0 +1,80 @@
+                return LocalDate.parse(str, DateTimeFormatter.ofPattern(s))
+                        .atStartOfDay();
+            } catch (DateTimeParseException e) {
+                //Fallthrough
```
Good!

anselmlong commented on src/main/java/mel/utils/Parser.java:
```diff
@@ -0,0 +1,80 @@
+     * @see LocalDateTime
+     */
+    public LocalDateTime parseDateTime(String str) throws ParseException {
+        for (String s : FORMAT_DATE) {
```
I like how you account for any date format!

devivy replied:
dedication

anselmlong commented on src/main/java/mel/utils/Storage.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Save data directory path.
+     */
+    private final File DIR_DATA = new File("data");
```
Good constant names.

anselmlong commented on src/main/java/mel/utils/UI.java:
Should be Ui.java!

anselmlong commented on src/test/java/mel/tasks/DeadlineTest.java:
```diff
@@ -0,0 +1,32 @@
+
+public class DeadlineTest {
+    @Test
+    public void constructor_invalidInput_exceptionThrown() {
```
Good naming!

anselmlong commented on text-ui-test/input.txt:
Comprehensive test!

anselmlong reviewed (commented):
Looks good to me! Some nits about naming and all, but everything looks generally good. 💯

kaixuan477 committed:
Update build.gradle for checkstyle

kaixuan477 committed:
Merge branch 'branch-A-CheckStyle'

kaikquah commented on src/main/java/mel/main/Mel.java:
```diff
@@ -0,0 +1,70 @@
+     * @param input task input string.
+     * @see TaskList
+     */
+    public void taskAction(String input) {
```
Possible to rename this method to make it sound more like a verb

kaikquah commented on src/main/java/mel/utils/Parser.java:
```diff
@@ -0,0 +1,80 @@
+     * Predetermined formats of date-time inputs,
+     * for parsing with DateTimeFormatter formatter.
+     */
+    private static final String[] FORMAT_DATE_TIME = {
```
I like how you included many different formats 👍

kaikquah commented on src/main/java/mel/utils/Ui.java:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Sends introductory text to user on Mel startup.
+     */
+    public void hello() {
```
similarly instead of naming it hello, you could use a verb in your phrases like greet or sayHello

kaikquah commented on src/test/java/mel/utils/ParserTest.java:
```diff
@@ -0,0 +1,35 @@
+
+public class ParserTest {
+    @Test
+    public void parseDateTime_validFormat_success() {
```
good naming and good test!

kaikquah commented on src/main/java/mel/utils/Ui.java:
```diff
@@ -0,0 +1,70 @@
+ * chatbot, to receive and respond to user input.
+ */
+public class Ui {
+    private static final String LOGO = " __    __        __  \n"
```
I like the use of a logo which I did not include in mine, haha

kaikquah reviewed (commented):
Overall I really like your code and am very impressed by it. Just some naming issues here and there. :)

kaixuan477 committed:
Update for GUI functionality

kaixuan477 committed:
Merge branch 'branch-Level-10'

kaixuan477 committed:
Update Mel.java

kaixuan477 committed:
Merge branch 'branch-A-Varargs'

devivy commented on src/main/java/mel/exceptions/MelException.java:
everything looks clean!

what's good:
- clear separation of exception types
- custom messages for each exception type
- consistent naming convention for exceptions

devivy commented on src/main/java/mel/exceptions/ParseException.java:
neat:)

what's good:

clear separation of exception types
custom messages for each exception type
consistent naming convention for exceptions

devivy commented on src/main/java/mel/exceptions/TaskException.java:
solid!!

what's good:
- clear separation of exception types
- custom messages for each exception type
- consistent naming convention for exceptions

devivy commented on src/main/java/mel/gui/Main.java:
everything looks clean! 😊
- clear separation of gui concerns from main logic
- use of fxml for ui layout

devivy commented on src/main/java/mel/tasks/Task.java:
perfect

devivy commented on src/test/java/mel/utils/ParserTest.java:
test case naming format followed:)

clean code!!

devivy commented on src/main/resources/images/mel.gif:
omg cuteeeeeeeeee

devivy commented on src/main/java/mel/utils/Ui.java:
ur so dedicated hahaha!!
such good code..
(maybe naming could be done better, more self explanatory and intuitive)

devivy reviewed (commented):
overall really really good code, well written, detailed and thoughtful:)

kaixuan477 committed:
Update GUI to intialise hello message

Mel chatbot GUI initialises without introductory message,
instead immediately awaits user input.

Common chatbot behaviour includes an introductory
message to signal chatbot startup completion, thus
ready for subsequent user input.

Updating Mel.java and MainWindows.java to include
hello message on startup is sufficient.

Dialogbox for hello message is generated on
startup, with a dedicated method for its function,
to adhere to code quality of SLAP.

kaixuan477 committed:
Update files to utilise java assertion

kaixuan477 committed:
Merge branch 'branch-A-Assertions'

kaixuan477 committed:
Update java files for code quality compliance

Java files lack compilance with code quality standards
covered, such as missing default branches
in case statements.

Updating java files respectively to ensure overall
code quality compilances are met.

kaixuan477 committed:
Merge branch 'master' into branch-A-CodeQuality

kaixuan477 committed:
Merge pull request #1 from kaixuan477/branch-A-CodeQuality

Compliance to code quality

kaixuan477 committed:
Set up CI for project workflow

kaixuan477 committed:
Update workflow for windows version runtest.bat only

kaixuan477 committed:
Update to implement MassOps

kaixuan477 committed:
Merge branch 'C-MassOps'

kaixuan477 committed:
Update GUI and layout

kaixuan477 committed:
Update files for error handling

kaixuan477 committed:
Update JUnit tests

kaixuan477 committed:
Add user guide in docs and update save file loading feature

kaixuan477 committed:
Update README.md

kaixuan477 committed:
Update README.md

kaixuan477 committed:
Update README.md
